tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc

```diff
@@ -1995,6 +2094,86 @@ void CopyTensorElements(const Tensor& tensor, protobuf::RepeatedField<T>* out) {
   std::copy(src, src + num_out_elements, dst);
 }
 
+template <DataType dtype, typename CType>
+void TestConvertVariableV2(VariableOpConverterTest* test) {
+  struct TestParam {
+    // TODO: more params
+    string container;
+    string shared_name;
+    std::vector<int> dims;
+    float epsilon;
+    Status conversion_status;
+  };
+
+  std::vector<TestParam> test_param = {
+      {"", "var0", {}, 0.001, Status::OK()},
+      {"", "var0", {64}, 0.001, Status::OK()},
+      {"", "var0", {8, 16}, 0.001, Status::OK()},
+      {"box", "var", {8, 16}, 0.001, Status::OK()}};
+  for (auto p : test_param) {
+    // Create node definition.
+    NodeDef node_def;
+    std::vector<int64_t> dims_64(p.dims.begin(), p.dims.end());
+    TensorShape shape = TensorShape(gtl::ArraySlice<int64_t>(dims_64));
+    TF_CHECK_OK(
+        NodeDefBuilder("my_var", "VariableV2")
+            .Attr("dtype", dtype)
+            .Attr("shape", shape)
+            .Attr("container", p.container)
+            .Attr("shared_name", p.shared_name)
+            .Finalize(&node_def));
+
+    OpKernel* kernel;
+    OpKernelContext* context;
+    test->CreateContext(node_def, &kernel, &context);
+
+    test->Reset(TrtPrecisionMode::FP32, TrtTestMode::kDynamicShape);
+
+    // Set the value of the variable according to p.dims.
+    int var_size = std::accumulate(p.dims.begin(), p.dims.end(), 1,
+                                   std::multiplies<int>());
+    std::vector<CType> expected_value;
+    expected_value.reserve(var_size);
+    for (int i = 0; i < var_size; i++) {
+      // Set expected_value[i] = (cast)i.
+      expected_value.push_back((CType)i);
+    }
+
+    // To set the variable, we get the tensor by executing the VariableV2 op
+    // rather than creating the resource directly in the manager, because:
+    // 1) LegacyVar defined in `variable_ops.cc` is not accessible.
+    // 2) Tensor::set_shape is private, VariableOp is a friend class.
+    kernel->Compute(context);
```

Comment:
This is to help me understand. This line execute the kernel, but the output doesn't have the value we want, so we copy the value to the output using cudaMemcpy below?

Reply:
Executing a `VariableV2` kernel gives you a tensor ref, however, the tensor isn't initialized until an `Assign` op is executed. The following lines do roughly what an assign op would do, that is to allocate the tensor and copy the value.